.travis.yml

```diff
@@ -1,52 +1,80 @@
 language: android
-sudo: false
+dist: precise
 android:
   components:
     # use the latest revision of Android SDK Tools
     - platform-tools
     - tools
-    - build-tools-25.0.3
-    - android-25
+    - build-tools-$ANDROID_BUILD_TOOLS
+    - android-22
+    - android-24
 
     # Additional components
+    - extra-google-google_play_services
+    - extra-google-m2repository
     - extra-android-m2repository
+    - addon-google_apis-google-19
 
     # Specify at least one system image,
     # if you need to run emulator(s) during your tests
+    - sys-img-armeabi-v7a-android-24
     - sys-img-armeabi-v7a-android-22
-jdk:
-  - oraclejdk8
 env:
   global:
-    - _FORCE_LOGS=1
     - DEVICE=android
+    - START_EMU=1
+    - ANDROID_EMU_NAME=test
+    - ANDROID_EMU_ABI=armeabi-v7a
+    - ANDROID_BUILD_TOOLS=26.0.2
+    - RECURSIVE=1
   matrix:
-    - TEST=unit RECURSIVE=1
-    - TEST=functional RECURSIVE=1
+    - TEST=unit START_EMU=0
+    - TEST=functional ANDROID_EMU_TARGET=android-24
 before_script:
-  # android stuff
-  - android list targets
-  - echo no | android create avd --force -n test -t android-21 --abi armeabi-v7a
-  - emulator -avd test -no-audio -no-window &
-  # skip waiting for start until after,
-  # since they can happen in parallel
-
   # node stuff
   - curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.33.1/install.sh | bash
   - nvm install 6
-  - npm install -g gulp
-  - npm install -g mocha
+  - node --version
+  - npm --version
+
+  # android stuff
+  - android list sdk --extended --no-ui --all
+  - android list targets
+
+  # npm stuff
   - npm install
 
-  # build the espresso server
-  - pushd espresso-server
-  - ./gradlew clean assembleAndroidTest
-  - popd
+  - |
+    if [ ${START_EMU} = "1" ]; then
+      echo y | android update sdk --no-ui -t platform-tools
+      echo y | android update sdk --no-ui -t tools
+      echo y | android update sdk --no-ui -t android-22
```

Comment:
We need the same fix here as for https://github.com/appium/appium-uiautomator2-server/pull/158 to save VM preparation time

Reply:
More than that can be done. Updated.

Reply:
👍

Reply:
@vmaxim FYI